### PR TITLE
[tvos] remove fork/spawnv for tvos platform

### PR DESCRIPTION
**Pull request by fuzzard**
TVOS cannot use fork or spawnv. Patch uses targetconditional.h to only affect
tvos platform and not ios.

### CODE REVIEW

File: depends/darwin_embedded/libarchive/03-android-include.patch [hunk 1]
```diff
@@ -0,0 +1,22 @@
+--- a/CMakeLists.txt
```
Comment (phunkyfish) [inline depends/darwin_embedded/libarchive/03-android-include.patch]: Do you need this patch file for tvos? Or is android in the title just misleading?

**Comment by fuzzard** [comment]
Honestly, I left it in because both windows leave it in. As far as I could tell, it's not useful for non android, but also doesn't affect any other platform if patched

**Comment by phunkyfish** [comment]
Cool

File: depends/darwin_embedded/libarchive/04-tvos-fork.patch [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
++
+ #if defined(HAVE_SYS_TYPES_H)
+ #  include <sys/types.h>
+ #endif
+@@ -148,7 +155,11 @@
+ #if HAVE_VFORK
+ 	child = vfork();
+ #else
+-	child = fork();
++#if TARGET_OS_TV
```
Comment (AlwinEsch) [inline depends/darwin_embedded/libarchive/04-tvos-fork.patch]: Have you looked why CMakeLists.txt on libarchive define one of this:
```
#if defined(HAVE_PIPE) && defined(HAVE_FCNTL) && \
    (defined(HAVE_FORK) || defined(HAVE_VFORK) || defined(HAVE_POSIX_SPAWNP))
```

There is a check about `HAVE_VFORK` and `HAVE_FORK` inside.
https://github.com/libarchive/libarchive/blob/614110e76d9dbb9ed3e159a71cbd75fa3b23efe3/CMakeLists.txt#L1302
https://github.com/libarchive/libarchive/blob/614110e76d9dbb9ed3e159a71cbd75fa3b23efe3/CMakeLists.txt#L1358

Comment (AlwinEsch) [reply depends/darwin_embedded/libarchive/04-tvos-fork.patch]: If not fixable by some values to his CMakeLists.txt try to add them in `vfs.libarchive/depends/common/` by a extra patch which only add the `#if TARGET_OS_TV` there, it seems no other differences between.

**Comment by AlwinEsch** [comment]
Be aware there comes also a update of them on https://github.com/xbmc/vfs.libarchive/pull/36 and seen now is also Libarchive 3.4.1 available.

If some change for TVOS still needed with 3.4.1 then I leave out the version update on the other request away and can be increased with them here,

**Comment by AlwinEsch** [comment]
As defined on slack.

Try the use of one of this patches (on `./vfs.libarchive/depends/common/libarchive`):
```diff
index 7e183b4..0f4d270 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1305,7 +1305,11 @@ CHECK_FUNCTION_EXISTS_GLIBC(fchmod HAVE_FCHMOD)
 CHECK_FUNCTION_EXISTS_GLIBC(fchown HAVE_FCHOWN)
 CHECK_FUNCTION_EXISTS_GLIBC(fcntl HAVE_FCNTL)
 CHECK_FUNCTION_EXISTS_GLIBC(fdopendir HAVE_FDOPENDIR)
-CHECK_FUNCTION_EXISTS_GLIBC(fork HAVE_FORK)
+IF(NOT TARGET_DARWIN_TVOS)
+  CHECK_FUNCTION_EXISTS_GLIBC(fork HAVE_FORK)
+ELSE()
+  SET(HAVE_FORK 0)
+ENDIF()
 CHECK_FUNCTION_EXISTS_GLIBC(fstat HAVE_FSTAT)
 CHECK_FUNCTION_EXISTS_GLIBC(fstatat HAVE_FSTATAT)
 CHECK_FUNCTION_EXISTS_GLIBC(fstatfs HAVE_FSTATFS)
@@ -1361,7 +1365,11 @@ CHECK_FUNCTION_EXISTS_GLIBC(unsetenv HAVE_UNSETENV)
 CHECK_FUNCTION_EXISTS_GLIBC(utime HAVE_UTIME)
 CHECK_FUNCTION_EXISTS_GLIBC(utimes HAVE_UTIMES)
 CHECK_FUNCTION_EXISTS_GLIBC(utimensat HAVE_UTIMENSAT)
-CHECK_FUNCTION_EXISTS_GLIBC(vfork HAVE_VFORK)
+IF(NOT TARGET_DARWIN_TVOS)
+  CHECK_FUNCTION_EXISTS_GLIBC(vfork HAVE_VFORK)
+ELSE()
+  SET(HAVE_VFORK 0)
+ENDIF()
 CHECK_FUNCTION_EXISTS_GLIBC(wcrtomb HAVE_WCRTOMB)
 CHECK_FUNCTION_EXISTS_GLIBC(wcscmp HAVE_WCSCMP)
 IF(NOT CMAKE_SYSTEM_NAME STREQUAL WindowsStore)
```
or
```diff
--- a/libarchive/filter_fork_posix.c
+++ b/libarchive/filter_fork_posix.c
@@ -32,6 +32,11 @@

 __FBSDID("$FreeBSD: head/lib/libarchive/filter_fork.c 182958 2008-09-12 05:33:00Z kientzle $");

+#ifdef TARGET_DARWIN_TVOS
+#undef HAVE_POSIX_SPAWNP
+#undef HAVE_VFORK
+#endif
+
 #if defined(HAVE_SYS_TYPES_H)
 #  include <sys/types.h>
 #endif
@@ -148,7 +153,11 @@ __archive_create_child(const char *cmd, int *child_stdin, int *child_stdout)
 #if HAVE_VFORK
        child = vfork();
 #else
-       child = fork();
+#ifdef TARGET_DARWIN_TVOS
+       child = -1;
+#else
+       child = fork();
+#endif
 #endif
        if (child == -1)
                goto stdout_opened;
```

**Comment by fuzzard** [comment]
sorry i havent got back to this. Im happy to wait until you do the version bumps and what not Alwin.

Having said that, whats the preference? Patch all systems that can via common, or to split patched into their own platform folder?

The cmake patch wont work, as the code in question does a check for vfork, else it runs fork(). fork is the actual problem for tvos (among a bunch of other things like spawn/spawnv, etc.)

Second patch should work, havent tested though. Its pretty much the equivalent of what ive PR'd, but i used the targetconditionals.h to do plstform checks for apple. my implementation couldnt go into common due to targetconditionals.h only being available on darwin platforms i believe.

Your's should work in common i think, as it just uses defines we set.

So, the question still stands, do you want patches all bundled into common, or split into their platform?

**Comment by fuzzard** [comment]
So, without being able to provide an optional define in common, your second option doesnt work either. I would need to be able to set -DTARGET_DARWIN_TVOS in flags.txt to be able to use that define in the code. 

~~is it possible to just add a flags.txt to depends/darwin_embedded and it uses the rest of the info from common? I would have thought not, but i definitely dont know the ins and outs of the addon build stuff, so correct me if im wrong.~~

We cant just blanket add the targetconditionals.h header, as far as im aware its only valid on apple builds.

Ive rebased to your latest updates, and have built locally all ok for tvos and ios (just to insure no side affect with changes)

**Comment by fuzzard** [comment]
@AlwinEsch Have added the ifndef AppleTV check as suggested.

**Comment by AlwinEsch** [comment]
@fuzzard OK to bring in and do my one after them?

**Comment by fuzzard** [comment]
Yeah, I think this is all good now

**Pull request merged**